lib/fetchContent.js

```diff
@@ -348,6 +348,227 @@ function getArticleRelations(days, facets, aspects)
+
+	if (!grouping[metadataKey1][term1.taxonomy][term1.name].hasOwnProperty(csvTerm2)) {
+		grouping[metadataKey1][term1.taxonomy][term1.name][csvTerm2] = 0;
+	}
+
+	grouping[metadataKey1][term1.taxonomy][term1.name][csvTerm2]++;
+}
+
+function aggregateArticles(articles, facets, minCorrelation, days, searchParams, filterByMetadataCsv=null)
+{
+	let articlesByUuid = {}; // UUID -> article
+	let articlesByFacetTaxonomyName = {}; // {metadataKey}{taxonomy}{name} -> [UUID]
+	let facetCorrelations = {}; // {metadataKey}{taxonomy}{name} {metadataKey:taxonomy:name} -> count
+
+	articles.forEach(article => {
+		let uuid = article.id;
+		let metadata = article.metadata;
+		let metadataKeys = Object.keys(metadata);
+
+		article.metadataCsvs = [];
+
+		// construct list of all facetTuples for this article
+		let facetTuples = []; // [ [metadataKey, term], ... ]
+		metadataKeys.forEach( metadataKey => {
+			if (metadataKey === 'primarySection' || metadataKey === 'primaryTheme') {
+				facetTuples.push({
+					metadataKey,
+					term : metadata[metadataKey].term
+				});
+			} else {
+				metadata[metadataKey].forEach( wrappedTerm => {
+					facetTuples.push({
+						metadataKey,
+						term : wrappedTerm.term
+					});
+				});
+			}
+		});
+
+		// construct the csv for each facetTuple
+		facetTuples.forEach( facetTuple => {
+			facetTuple.csv = [facetTuple.metadataKey, facetTuple.term.taxonomy, facetTuple.term.name].join(':');
+		});
+
+		// check if we have the specified filterByMetadataCsv (if not null),
+		if (filterByMetadataCsv !== null) {
+			let tuplesMatchingFilter = facetTuples.filter( facetTuple => facetTuple.csv === filterByMetadataCsv );
+			if (tuplesMatchingFilter.length == 0) {
+				return; // ignore this article if no tuples match
+			}
+		}
+
+		articlesByUuid[uuid] = article;
+
+		// filter out people who are authors
+		let knownAuthors = {};
+		facetTuples.forEach( facetTuple => {
+			if (facetTuple.term.taxonomy == 'authors') {
+				knownAuthors[facetTuple.term.name] = true;
+			}
+		})
+
+		facetTuples = facetTuples.filter( facetTuple => facetTuple.term.taxonomy !== 'people' || ! knownAuthors[facetTuple.term.name]);
+
+		// construct articlesByFacetTaxonomyName
+		facetTuples.forEach( facetTuple => {
+			addUuidToMetadataTerm(articlesByFacetTaxonomyName, facetTuple.metadataKey, facetTuple.term, uuid);
+		});
+
+		// construct facetCorrelations
+		facetTuples.forEach( facetTuple1 => {
+			facetTuples.forEach( facetTuple2 => {
+				if ( facetTuple1.csv !== facetTuple2.csv ) {
+						addFacetCorrelation( facetCorrelations, facetTuple1.metadataKey, facetTuple1.term,
+																										facetTuple2.metadataKey, facetTuple2.term);
+					}
+			});
+		});
+
+		// refactor metadata into CSVs
+		facetTuples.forEach( facetTuple => {
+			article.metadataCsvs.push( facetTuple.csv );
+		});
+	});
+
+	// strip out weak correlations, using minCorrelation
+
+	Object.keys( facetCorrelations ).forEach(metadataKey => {
```

Comment:
The 4 loops make me dizzy.

Reply:
Is necessary if we stick with the nested {metadataKey}{taxonomy}{name} structure, but not if we switch completely to metadataKey:taxonomy:name.
Still thinking this through.

Reply:
A side-effect is that when the final metadataKey:taxonomy:name map is constructed, it is in a sensible order, which would not be the case if we skipped the {metadataKey}{taxonomy}{name} stage.

Reply:
possible refactor for later